util/chplenv/chplenv_verify.py

```diff
@@ -0,0 +1,289 @@
+"""
+Verifies that the current environment is set up correctly for building Chapel, and that once built the compiler will function correctly.
+"""
+
+import functools
+from utils import try_run_command, memoize
+import tempfile
+import shutil
+import sys
+import os
+import shlex
+
+passes = []
+
+def log(*args, **kwargs):
+    kwargs = {**kwargs, "file": sys.stderr}
+    verbose = False
+    if "verbose" in kwargs:
+        verbose = kwargs["verbose"]
+        del kwargs["verbose"]
+    if verbose:
+        print(*args, **kwargs)
+
+def register(_cls=None, *, after=None):
+    def decorator_register(cls):
+        global passes
+        if after:
+            idx = next(
+                i
+                for i, (name, _) in enumerate(passes)
+                if name == after.__name__
+            )
+            passes.insert(idx + 1, (cls.__name__, cls))
+        else:
+            passes.append((cls.__name__, cls))
+
+        @functools.wraps(cls)
+        def wrapper_register(*args, **kwargs):
+            return cls(*args, **kwargs)
+
+        return wrapper_register
+
+    if _cls is None:
+        return decorator_register
+    else:
+        return decorator_register(_cls)
```

Comment:
in this case, I suspect the decorator hurts rather than helps. You're maintaining a list (and inserting into that list). You use names to find places in that list to insert. Also, all of your classes are defined such that earlier verification steps occur first.

Why not just have a list literal of all the steps somewhere at the bottom? The net result will be the same (you have a list of checks to run), and it would actually be a bit easier to reason about for the reader of this file, because instead of having to keep track of where each insertion places each test, you can just read the list and know what order the checks will occur.

Reply:
My goal here was to avoid having to specify stuff twice, but I take your point. I will use a global list

Reply:
you could have some kind of introspetion find all classes in the current namespace, too, if you wanted.

Reply:
I am not sure how that provides any benefit over the attributes approach. If the concern is "its hard to understand the exact order these tests run in", I don't see how metaclasses could solve that. For introspection, maybe its the order they are defined. But that feels more subtle and brittle than using attributes/metaclasses.